File: 7-kyu-kata/nth-floyd-line/python/nth_floyd_line_test.py
# 7 kyu - nth Floyd line  [ ID: 5b096efeaf15bef812000010  (nth-floyd-line) ]
# URL: https://www.codewars.com/kata/5b096efeaf15bef812000010
# Category: REFERENCE  |  Tags: MATHEMATICS | PUZZLES | FUNDAMENTALS
# ******************************************************************************
import random as r


def ktest():
    def static():

        testc = [
            [15, 5],
            [26, 7],
            [17, 6],
            [24, 7],
            [19, 6],
            [5, 3],
            [212, 21],
        ]
        r.shuffle(testc)

        test.it("Static tests")
        for n, s in testc:
            soln = s
            test.assert_equals(nth_floyd(n), soln, "Wrong Value")

    static()

    def nf(n, l=0, r=10e8):
        strt = lambda l: int((1 + (l - 1) * l) - (((l - 1) / 2) * (2 + (l - 2))))
        while l <= r:
            m = int(l + r) // 2
            begin = strt(m)
            end = begin + m - 1
            if begin <= n and end >= n:
                return m
            elif end < n:
                l = m + 1
            elif begin > n:
                r = m - 1

    def rand():
        test.it("Random tests")
        for i in range(100):
            n = r.randint(1, 10**5)
            soln = nf(n)
            test.assert_equals(nth_floyd(n), soln, "Wrong value")

        for i in range(100):
            n = r.randint(10**5, 10**9)
            soln = nf(n)
            test.assert_equals(nth_floyd(n), soln, "Wrong value")

    rand()


ktest()

File: 7_Kyu_Kata/nth-floyd-line/JavaScript/nthFloydLine.Test.js
// [object Object] - nth Floyd line  [ ID: 5b096efeaf15bef812000010  (nth-floyd-line) ]
// URL: https://www.codewars.com/kata/5b096efeaf15bef812000010
// Category: REFERENCE  |  Tags: MATHEMATICS | PUZZLES | FUNDAMENTALS
// *****************************************************************************
const _ = require ('lodash')

describe("Basic tests", function(){
Test.assertEquals(nthFloyd(15),5);
Test.assertEquals(nthFloyd(26),7);
Test.assertEquals(nthFloyd(17),6);
Test.assertEquals(nthFloyd(24),7);
Test.assertEquals(nthFloyd(19),6);
Test.assertEquals(nthFloyd(5),3);
Test.assertEquals(nthFloyd(212),21);
Test.assertEquals(nthFloyd(499502),1000);
});

describe("Random tests", function(){
for (let i = 0; i < 100; i++) {
    let r = _.random(1,Math.pow(10,5));
    let exp = kj6tr(r);        
    Test.assertEquals(nthFloyd(r),exp);
    r = _.random(Math.pow(10,5),Math.pow(10,9));
    exp = kj6tr(r);        
    Test.assertEquals(nthFloyd(r),exp);   
 }  
})

function kj6tr(n){
  return Math.floor(((1+8*(n-1))**0.5+1)/2)
}
